models/demos/t3000/llama2_70b/tt/llama_rope.py

```diff
@@ -0,0 +1,143 @@
+        self.device = device
+        self.is_mesh_device = isinstance(device, ttnn._ttnn.multi_device.MeshDevice)
+
+        self.core_grid = device.compute_with_storage_grid_size()
+        num_cores = self.core_grid.x * self.core_grid.y
+
+        # Generate the cos/sin matrices needed for ttnn.embedding op
+        cos_matrix, sin_matrix = compute_gather_cos_sin(
+            dhead=head_dim,
+            end=max_seq_len * 2,
+            theta=rope_theta,
+            position_ids=torch.arange(max_seq_len),
+            use_scaled_rope=use_scaled_rope,
+        )
+
+        self.cos_matrix = ttnn.from_torch(
+            cos_matrix,
+            device=device,
+            layout=ttnn.ROW_MAJOR_LAYOUT,
+            dtype=datatype,
+            mesh_mapper=ReplicateTensorToMesh(device) if self.is_mesh_device else None,
+        )
+        self.sin_matrix = ttnn.from_torch(
+            sin_matrix,
+            device=device,
+            layout=ttnn.ROW_MAJOR_LAYOUT,
+            dtype=datatype,
+            mesh_mapper=ReplicateTensorToMesh(device) if self.is_mesh_device else None,
+        )
+
+        # Generate the transformation matrix
+        trans_mat = get_rot_transformation_mat(dhead=ttnn.TILE_SIZE).repeat(
+            1, 1, num_cores, 1
+        )  # Repeat across all cores on device
+        trans_mat_mem_config = ttnn.create_sharded_memory_config(
+            shape=(1, 1, ttnn.TILE_SIZE * num_cores, ttnn.TILE_SIZE),
+            core_grid=ttnn.CoreGrid(y=self.core_grid.y, x=self.core_grid.x),
+            strategy=ttnn.ShardStrategy.HEIGHT,
+            orientation=ttnn.ShardOrientation.ROW_MAJOR,
+        )
+        self.transformation_mat = ttnn.from_torch(
+            trans_mat,
+            device=device,
+            layout=ttnn.TILE_LAYOUT,
+            dtype=datatype,
+            memory_config=trans_mat_mem_config,
+            mesh_mapper=ReplicateTensorToMesh(device) if self.is_mesh_device else None,
+        )
+
+    def get_trans_mats(self):
+        assert self.transformation_mat is not None, "Transformation matrix not initialized"
+        return self.transformation_mat
+
+    def get_rot_idxs(self, position_idxs):
+        assert isinstance(position_idxs, torch.Tensor), "Position ids must be a torch tensor"
+
+        batch = position_idxs.shape[0]
+        position_idxs = position_idxs.unsqueeze(0)
+        assert position_idxs.shape == (1, batch), "position idxs must be a [1, batch] tensor"
+        assert torch.min(position_idxs) >= 0, "position idxs must be non-negative"
+
+        rot_idxs = ttnn.as_tensor(
+            position_idxs,
+            dtype=ttnn.uint32,
+            layout=ttnn.ROW_MAJOR_LAYOUT,
+            mesh_mapper=ReplicateTensorToMesh(self.device) if self.is_mesh_device else None,
+        )
+
+        return rot_idxs
+
+    def get_rot_mats(self, position_idxs, device=None, return_rot_idxs=False):
+        device = self.device if device is None else device
```

Comment:
what is the use case for passing a device to this function which is not the device this class was created with?